Lib/fontTools/designspaceLib/__init__.py

```diff
@@ -434,27 +440,37 @@ def _addRule(self, ruleObject):
         # if none of the conditions have minimum or maximum values, do not add the rule.
         self.rules.append(ruleObject)
         ruleElement  = ET.Element('rule')
-        ruleElement.attrib['name'] = ruleObject.name
-        for cond in ruleObject.conditions:
-            if cond.get('minimum') is None and cond.get('maximum') is None:
-                # neither is defined, don't add this condition
-                continue
-            conditionElement = ET.Element('condition')
-            conditionElement.attrib['name'] = cond.get('name')
-            if cond.get('minimum') is not None:
-                conditionElement.attrib['minimum'] = self.intOrFloat(cond.get('minimum'))
-            if cond.get('maximum') is not None:
-                conditionElement.attrib['maximum'] = self.intOrFloat(cond.get('maximum'))
-            ruleElement.append(conditionElement)
+        if ruleObject.name is not None:
+            ruleElement.attrib['name'] = ruleObject.name
+        for conditions in ruleObject.conditionSets:
+            conditionsetElement = ET.Element('conditionset')
+            for cond in conditions:
+                if cond.get('minimum') is None and cond.get('maximum') is None:
+                    # neither is defined, don't add this condition
+                    continue
+                conditionElement = ET.Element('condition')
+                conditionElement.attrib['name'] = cond.get('name')
+                if cond.get('minimum') is not None:
+                    conditionElement.attrib['minimum'] = self.intOrFloat(cond.get('minimum'))
+                if cond.get('maximum') is not None:
+                    conditionElement.attrib['maximum'] = self.intOrFloat(cond.get('maximum'))
+                conditionsetElement.append(conditionElement)
+            if len(conditionsetElement):
+                ruleElement.append(conditionsetElement)
+        # XXX shouldn't we require at least one sub element?
+        # if not ruleObject.subs:
+        #     raise DesignSpaceDocument('Invalid empty rule with no "sub" elements')
```

Comment:
What's the use of a rule that doesn't specify any substitutions? Is it the equivalent of a "ignore sub" in feature files?

Reply:
No, it does not mean `ignore sub`. Would that be useful here?

Reply:
Nah. An empty subs attr _could_ mean: "this is work in progress, I will add substitutions later"

Reply:
ok fine, will remove the `# XXX` lines

Reply:
in https://github.com/fonttools/fonttools/blob/master/Doc/source/designspaceLib/readme.rst#513-sub-element, it says about `sub` that

> This element is optional. It may be useful for editors to know which glyphs can be used to preview the axis.

I understand that one may want to write incomplete data, and that a rule without substitutions could simply be ignored when the designspace document is processed by a UFO or varfont builder. However, stressing the fact that `sub` element is "optional" seems counter-intuitive to me.
The sub element is not only "useful for editors to know which glyphs to preview".

Reply:
It's possibly a remnant of what @LettError originally thought the `<rule>` element was going to be for? Definitely needs rephrasing.

Reply:
Yes, it should be rephrased. Or maybe it is enough to remove that line.

Reply:
Here are some changes to the designspaceLib/readme.rst https://github.com/fonttools/fonttools/commit/d26bb392512b54444ca879b0d09c8f15333b78e6
These are the `conditionsets`, `conditions` and `sub` element changes discussed here.

Reply:
https://github.com/fonttools/fonttools/commit/d26bb392512b54444ca879b0d09c8f15333b78e6